### PR TITLE
feat: add plumbing for `pathfinder_getProof`

### LINKED ISSUE #778: Add framework for `pathfinder_` methods
This is a followup work needed on #726 

See https://github.com/eqlabs/pathfinder/pull/726#discussion_r1043235972

### PR DESCRIPTION
**Pull request by Mirko-von-Leipzig**
This PR adds the final bit of plumbing and documentation required by `pathfinder_getProof`.

Closes #778.

### Refactoring

This PR also refactors the RPC context and module system slightly. Since the pathfinder extension methods now also want access to context and the error framework, I've moved and slightly reworked the context and module code.

We now have our `Module` builder system which encapsulates method registration.

I have still left v0.1 as is because I'm scared of touching it :D

### get proof

I've added `pathfinder_getProof` to both pathfinder and starknet v0.2 endpoints.

I've also attempted to document the get proof method in the json rpc spec, but I struggled -- I'm fairly sure I've done something incorrect here. It doesn't display correctly on the playground (specifically the array of nodes).

I'd appreciate some help with double checking my types and outputs especially.

### CODE REVIEW

File: crates/rpc/src/module.rs [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+            .address(addr)
+            .build()
+            .unwrap();
+
+        let message = client
+            .request::<String>("echo", rpc_params!["testing testing 123"])
+            .await
+            .unwrap();
+        assert_eq!(message, "testing testing 123");
+    }
```
Comment (Mirko-von-Leipzig) [inline crates/rpc/src/module.rs:181]: @CHr15F0x as mentioned offline, I can't get this test to work correctly. I'd appreciate some kung-fu here.

Comment (Mirko-von-Leipzig) [reply crates/rpc/src/module.rs]: Thanks 🙏

File: README.md [hunk 2]
```diff
@@ -332,11 +332,16 @@ To check if it's running well use `docker-compose logs -f`.
 
 ## JSON-RPC API
 
-Pathfinder supports both version `v0.1.0` and `v0.2.1` of the Starknet JSON-RPC specification.
+You can get interact with StarkNet using the JSON-RPC API. Pathfinder supports the official StarkNet RPC API and in addition supplements this with its own pathfinder specific extensions such as `pathfinder_getProof`.
```
Comment (CHr15F0x) [inline README.md:335]: ?

```suggestion
You can interact with StarkNet using the JSON-RPC API. Pathfinder supports the official StarkNet RPC API and in addition supplements this with its own pathfinder specific extensions such as `pathfinder_getProof`.
```

File: crates/pathfinder/rpc_pathfinder_examples.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+          -d "${1}" \
+          http://127.0.0.1:9545
+     printf "\n\n"
+}
+
+rpc_call '{"jsonrpc":"2.0","id":"0","method":"pathfinder_version"}'
+
+rpc_call '{
+    "jsonrpc": "2.0",
+    "method": "starknet_addInvokeTransaction",
```
Comment (CHr15F0x) [inline crates/pathfinder/rpc_pathfinder_examples.sh:18]: `pathfinder_getProof` ?

File: README.md (hunk 2 above)
Comment (Mirko-von-Leipzig) [reply README.md]: Thanks!

**Review by CHr15F0x — APPROVED** [review]
lgtm :+1:

**Pull request merged**